cli/azd/pkg/tools/github/github.go

```diff
@@ -33,10 +48,75 @@ type GitHubCli interface {
 	GitHubActionsExists(ctx context.Context, repoSlug string) (bool, error)
 }
 
-func NewGitHubCli(commandRunner exec.CommandRunner) GitHubCli {
-	return &ghCli{
+func NewGitHubCli(ctx context.Context, console input.Console, commandRunner exec.CommandRunner) (GitHubCli, error) {
+	return newGitHubCliImplementation(ctx, console, commandRunner, http.DefaultClient, downloadGh, extractGhCli)
+}
+
+// cGitHubCliVersion is the minimum version of GitHub cli that we require (and the one we fetch when we fetch bicep on
+// behalf of a user).
+var cGitHubCliVersion semver.Version = semver.MustParse("2.22.1")
+
+// newGitHubCliImplementation is like NewGitHubCli but allows providing a custom transport to use when downloading the
+// GitHub CLI, for testing purposes.
+func newGitHubCliImplementation(
+	ctx context.Context,
+	console input.Console,
+	commandRunner exec.CommandRunner,
+	transporter policy.Transporter,
+	acquireGitHubCliImpl GetGitHubCliImplementation,
+	extractImplementation ExtractGitHubCliFromFileImplementation,
+) (GitHubCli, error) {
+	if override := os.Getenv("AZD_GH_CLI_TOOL_PATH"); override != "" {
+		log.Printf("using external github cli tool: %s", override)
+
+		return &ghCli{
+			path:          override,
+			commandRunner: commandRunner,
+		}, nil
+	}
+
+	githubCliPath, err := azdGithubCliPath()
+	if err != nil {
+		return nil, fmt.Errorf("finding github cli: %w", err)
+	}
+
+	if _, err = os.Stat(githubCliPath); err != nil && !errors.Is(err, os.ErrNotExist) {
+		return nil, fmt.Errorf("finding github cli: %w", err)
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		if err := os.MkdirAll(filepath.Dir(githubCliPath), osutil.PermissionDirectory); err != nil {
+			return nil, fmt.Errorf("downloading github cli: %w", err)
+		}
+
+		msg := "Downloading Github cli"
+		console.ShowSpinner(ctx, msg, input.Step)
+		err = acquireGitHubCliImpl(ctx, transporter, cGitHubCliVersion, extractImplementation, githubCliPath)
```

Comment:
@savannahostrowski / @danieljurek Do we run afoul of any notice agreements if we do not supply user notice as part of this install? With `bicep` it was all Microsoft, and with `gh` we do have any extra company boundary.

My preference would be to prompt the user if they wanted to install. It would also be preferable to use any official installation mechanisms if possible to avoid any changes there that would impact us.

Reply:
I left some comments about this experience in the demo chat. One of them was around needing an extra prompt for user consent.

Reply:
I'm hoping to `not prompt` user.

My dream is to sell this extra binary as if it is part of `azd` directly.  We are just not adding the code or the bits inside the `azd binary` directly because not everyone might want to use `github repos/actions`.

We can even get and save the binary with a name like `azd-gh` or something which gives a hint about this is expected to be part of `azd files`.

Actually, we can build our own gh-cli and put it in a repo where we get it from and only for `azd`.

So, for uses, I don't want azd to say `we are installing gh cli`, but we are getting an extra peace of azd to handle github requests.  Does it make sense?

Reply:
@vhvb1989 I'm speaking purely from a privacy standpoint. I understand where you're coming from, from an engineering perspective. If there's no privacy requirement, I'm happy with the changes as-is.

Reply:
So I have two concerns with this approach:

1. The consent aspect - it feels a little jarring to see a message that says  "Installing the GitHub CLI" after accepting a prompt that didn't articulate that this will happen. In all cases like this, I would encourage us to get explicit consent.

3. The user now having multiple "copies" of the CLI on their machine. If I see "installing GitHub CLI" as part of this process, I might be inclined to think I have the GitHub CLI globally available, which is not true with this design. This could lead to users thinking we are doing something nefarious or "magical".  To @bwateratmsft's point in our team chat, it does seem like a more consistent UX if we check PATH to see if it exists, if it's not there, ask if they want to install and then do a real install so it is globally available. I think this is still a streamlined user experience and eliminates the overhead of having a user go search for the right install command.

Reply:
>  it feels a little jarring to see a message that says "Installing the GitHub CLI"

We are not printing such message. We are using `Downloading Github cli`.  And the message is not persisted in the log. It actually takes about ~1 sec (the gh cli is really small).

> The user now having multiple "copies" of the CLI on their machine

There's a similar behavior when we install Visual Studio. For C++, for example, it brings `msvc` compiler,  `cmake build tools`, etc.  And those tools are inside the Visual Studio installation folder.  This way, Visual Studio is 100% sure about those versions are fully compatible.

Just imagine the next flow (where we would be trying to be aware of a current installation):

- azd checks if gh-cli is already installed in the system
  - A version `9.9` is found, but `azd` was tested on CI with version `8.0`.  Should we:
    - Show a warning about user has an untested version and it might work or it might not?
    - Ask user to downgrade version before using azd?
    - Don't say anything and take the risk of failing (because there's a breaking change in version `9.9`)
    - Ask user if they want azd to install version `8.0` , regardless of what version the user already has?

The same sort of questions would come if the version in the system is older than what azd was tested during CI.

In the current approach, we are *only* setting a minimum version to be installed for azd to work, but we are not 100% sure about all the available versions since the Min-version to the Last-version are fully compatible with azd.

Reply:
Okay, I spent the evening thinking about this some more. Here's what I'd recommend:

1. I don't think that we should at all surface that we are downloading the gh CLI, even if it's ephemeral. We aren't doing that for the Bicep CLI, which I believe is used in a similar way under the hood. Really the gh CLI is an implementation detail and unless we get feedback that this takes a long time, I don't think it's a consistent UX to print it out.
2. As long as we are sure to remove all deps (e.g. gh CLI) on uninstall, I am fine with this pattern of only having it as a "component" in azd (and not globally available).

Does that make sense?

Reply:
Agreed @savannahostrowski 

1. Regarding pinning vs not-pinning to specific version.
   - For GH CLI, we pin it specifically. Since our deps on the GH CLI login features is deterministic.
   - For Bicep CLI, we pick the latest I believe and should not pin to specific. Since, the features evolve on Bicep side and we should let it flow?

2. Uninstall. I don't think we remove the downloaded bits. Since its part of .azd/bin config?

@ellismg @vhvb1989

Reply:
@rajeshkamal5050 , we are targeting bicep version `0.12.40` . So, we don't target last version (which is currently `14.46`)

I am trying to convince @danieljurek to either automatically clean the ~/.azd folder content, or to at least prompt user and ask if they want to keep application data. But, for now, we are leaking the app data after uninstalling.

Reply:
> we are targeting bicep version 0.12.40 . So, we don't target last version (which is currently 14.46)

Shouldn't we rather not pin to a specific version for Bicep. Since its more of a functionality exposed directly to our end users provisioning features. Pinning, would mean we Azd is blocking the bicep features flowing through? @ellismg

Reply:
The good thing about pinning `bicep version` is to keep a more stable CI pipelines.

If we switch to always pick `latest`, our CI pipelines would scream at any breaking change from `bicep` (or regression).

We do `latest` for some tools in CI, like `golangci-lint`, which break our builds from time to time, as we try to keep compatible with latest version there.

@ellismg , just in case we would want to go `latest version` for bicep, we can use the magic url `https://github.com/Azure/bicep/releases/latest/download/<binary-name>` to fetch latest.   

However, I would vote to keep bicep and gh-cli with a manual switch to upgrade versions.  I don't like those moments when builds fails and the time it take to figure out there's a break coming from a dependency update.

Reply:
If our goal is to ensure that the `gh` CLI is available then maybe we should check for it on install and require that the user to install it if it's not present? ([install instructions for Linux](https://github.com/cli/cli/blob/trunk/docs/install_linux.md) differ based on distro) In the case of package managers, we can probably make it a dependency and the user will install it transparently. 

A few things to consider w/r/t cleaning out the user's `~/.azd` folder: 

In some cases we've seen users uninstall and reinstall `azd` on their machine to repair an installation. This can happen if you're downgrading from `daily` to `latest` builds or transitioning from a pre-MSI build to an MSI build (though that bug should be fixed in the latest release). In these cases, I don't think a user would want to reconstruct their `~/.azd` folder. That is, not all uninstall scenarios should delete user data. 

Per-machine installs mean that any user could have a `~/.azd` folder. We'd have to traverse through all of them and try to remove files. Per-machine installs happen on Linux/macOS and are possible on Windows. 

If we scoped this just to deleting `~/.azd/bin` folder that might be better than erasing user data... but we'd still be taking considerable liberties doing that in per-machine scenarios.

Reply:
> This can happen if you're downgrading from daily to latest builds 

The same app-config is not warranted to work for a previous or newer version. Azd can change the structure of the config files at any moment.  
The only valid scenario is uninstalling and reinstalling the same version of `azd`. But for a version switch, the ~/.azd folder should be restored.

Reply:
@vhvb1989 What is the final conclusion here regarding uninstall? Do we have something logged for removing `~/.azd/bin`? Or have we addressed @savannahostrowski's concern differently?

Reply:
The `uninstall` scripts are not deleting the `user-settings` folder.

Filed a new issue to define what we want there: https://github.com/Azure/azure-dev/issues/1569